### PR TITLE
Add initial structs to generate crds for ISM Policies.

### PR DESCRIPTION
**Pull request by IshaGirdhar**

### CODE REVIEW

**Comment by IshaGirdhar** [comment]
@swoehrl-mw I am working on the controller and reconciler, meanwhile wanted to have your eyes on the initial structs for crd generation. I have kept the policy_id for now for the override scenario, its optional and if not provided then will use metadata.name.

**Review by swoehrl-mw — COMMENTED** [review]
@IshaGirdhar Looks mostly fine to me.
Please keep in mind that we can only merge complete features, as such I'd like to keep this PR in draft mode to avoid accidental merge.

**Review by idanl21 — COMMENTED** [review]
Hey @IshaGirdhar, looks good
Actually the docs and the UserGuide is missing for me, i understood the main idea of the ISM but i didnt understood how it will be implemented and hows the configuration will look like at the end.
I think that it will be good to see it before continue so we can be all aligned about the feature.

File: opensearch-operator/api/v1/opensearchism_types.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+	Spec              ISMPolicySpec `json:"spec,omitempty"`
+}
+
+// ISMPolicySpec is the specification for the ISM policy for OS.
+type ISMPolicySpec struct {
+	// The default starting state for each index that uses this policy.
+	DefaultState string `json:"default_state"`
+	// A human-readable description of the policy.
+	Description       string            `json:"description"`
+	ErrorNotification ErrorNotification `json:"error_notification,omitempty"`
```
[the remark this answers is not in the document]
Comment (IshaGirdhar) [reply opensearch-operator/api/v1/opensearchism_types.go:24]: cool will fix it.

**Comment by IshaGirdhar** [comment]
> @IshaGirdhar Looks mostly fine to me. Please keep in mind that we can only merge complete features, as such I'd like to keep this PR in draft mode to avoid accidental merge.

Sure. Cool thanks.

**Comment by IshaGirdhar** [comment]
> Hey @IshaGirdhar, looks good Actually the docs and the UserGuide is missing for me, i understood the main idea of the ISM but i didnt understood how it will be implemented and hows the configuration will look like at the end. I think that it will be good to see it before continue so we can be all aligned about the feature.

Sure Let me come up with a draft UserGuide  and some documentation as well before moving ahead.

**Comment by IshaGirdhar** [comment]
@idanl21 
This implementation introduces a Custom Resource Definition (CRD) to represent an ISM (Index State Management) policy. Each ISM policy will be associated with a specific OS (OpenSearch) cluster. The process begins with the user creating a CR (Custom Resource), which triggers a controller written specifically to handle ISM. The controller will call the ISM policy reconciler, monitoring the CR for creation, updation, or deletions.

When the CR is created, updated, or deleted, the controller seamlessly connects to the OS cluster and executes the appropriate ISM policy APIs provided by OpenSearch. This ensures that the ISM policy is applied or removed correctly based on the user's actions.

Furthermore, users have the flexibility to specify multiple index patterns, allowing them to apply the policy to multiple indices that match the given pattern. Similarly, if a user wishes to remove the policy from specific indices following a certain pattern, they can simply remove that pattern from the policy.

A sample policy CR looks like this
```
apiVersion: opensearch.opster.io/v1
kind: ISMPolicy
metadata:
  name: example-policy
spec:
  opensearchCluster:
    name: my-first-cluster
  description: Example policy
  policy_id: example-policy
  default_state: hot
  states:
    - name: hot
      actions:
        - replica_count:
            number_of_replicas: 4
      transitions:
        - state_name: warm
          conditions:
             min_index_age: "10d"
    - name: warm
      actions:
        - replica_count:
            number_of_replicas: 2
      transitions:
        - state_name: delete
          conditions:
            min_index_age: "30d"
    - name: delete
      actions:
        - name : delete
          delete: {}
  ism_template:
    index_patterns:
      - "logs-*"
    priority: 5
   ```


This will create an ISM policy in the OS cluster
Let me know if you are looking for some more information.

**Comment by swoehrl-mw** [comment]
Superseded by #575

**Pull request closed without merging**